### PR TITLE
first

### PR DESCRIPTION
**Pull request by voloshynamaryna11**

### CODE REVIEW

File: src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+            throw new DataProcessingException("Can`t get orders of user with id = " + userId, e);
+        }
+        for (Order order : orderList) {
+            order.setProducts(getProductsOfOrder(order.getId()));
+        }
+        return orderList;
+    }
+
+    @Override
+    public Order create(Order item) {
```
Comment (KostiantynVorobiov) [inline src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java:42]:
```suggestion
    public Order create(Order order) {
```

File: src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+            statement.setLong(1, orderId);
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                Long productId = resultSet.getLong("id");
+                String productName = resultSet.getString("name");
+                double productPrice = resultSet.getDouble("price");
+                Product product = new Product(productName, productPrice);
+                product.setId(productId);
+                products.add(product);
+            }
```
Comment (KostiantynVorobiov) [inline src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java:156]: нужно закрывать statement, что бы не было открытых statement-ов в statement-ах. Во всех подобных методах

File: src/main/java/internet/market/dao/jdbc/UserDaoJdbcImpl.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+                    + login, e);
+        }
+        if (user != null) {
+            user.setRoles(getRolesOfUser(user.getId()));
+        }
+        return Optional.ofNullable(user);
+    }
+
+    @Override
+    public User create(User item) {
```
Comment (KostiantynVorobiov) [inline src/main/java/internet/market/dao/jdbc/UserDaoJdbcImpl.java:44]:
```suggestion
    public User create(User user) {
```

File: src/main/java/internet/market/dao/jdbc/ShoppingCartDaoJdbcImpl.java [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        }
+    }
+
+    @Override
+    public boolean delete(ShoppingCart shoppingCart) {
+        return delete(shoppingCart.getId());
+    }
+
+    @Override
+    public ShoppingCart create(ShoppingCart item) {
```
Comment (KostiantynVorobiov) [inline src/main/java/internet/market/dao/jdbc/ShoppingCartDaoJdbcImpl.java:60]:
```suggestion
    public ShoppingCart create(ShoppingCart shoppingCart) {
```

File: src/main/java/internet/market/dao/jdbc/ShoppingCartDaoJdbcImpl.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+            throw new DataProcessingException("Can't get all shoppingCarts", e);
+        }
+        for (ShoppingCart cart : allCarts) {
+            cart.setProducts(getProductsFromCart(cart.getId()));
+        }
+        return allCarts;
+    }
+
+    @Override
+    public ShoppingCart update(ShoppingCart item) {
```
Comment (KostiantynVorobiov) [inline src/main/java/internet/market/dao/jdbc/ShoppingCartDaoJdbcImpl.java:119]: same here

File: src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+            throw new DataProcessingException("Can't get all orders", e);
+        }
+        for (Order order : allOrders) {
+            order.setProducts(getProductsOfOrder(order.getId()));
+        }
+        return allOrders;
+    }
+
+    @Override
+    public Order update(Order item) {
```
Comment (KostiantynVorobiov) [inline src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java:101]: rename item to order

File: src/main/java/internet/market/dao/jdbc/UserDaoJdbcImpl.java [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+        }
+
+        for (User user : userList) {
+            user.setRoles(getRolesOfUser(user.getId()));
+        }
+        return userList;
+    }
+
+    @Override
+    public User update(User item) {
```
Comment (KostiantynVorobiov) [inline src/main/java/internet/market/dao/jdbc/UserDaoJdbcImpl.java:107]: same here

**Review by KostiantynVorobiov — CHANGES_REQUESTED** [review]
ОК!)

File: src/main/resources/init_db.sql [hunk 8]
```diff
@@ -1,8 +1,102 @@
...
+    REFERENCES `internet_shop`.`products` (`product_id`)
+    ON DELETE NO ACTION
+    ON UPDATE NO ACTION);
+
+ALTER TABLE `internet_shop`.`users`
+ADD COLUMN `user_login` VARCHAR(255) NOT NULL AFTER `deleted`;
+
+ALTER TABLE `internet_shop`.`users`
+DROP COLUMN `product_price`,
+ADD COLUMN `user_password` VARCHAR(255) NOT NULL AFTER `user_login`;
```
Comment (KseniiaMakarova) [inline src/main/resources/init_db.sql:98]: let's better refactor the CREATE TABLE statement

File: src/main/java/internet/market/dao/jdbc/UserDaoJdbcImpl.java [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+            statement.setString(1, item.getName());
+            statement.setString(2, item.getLogin());
+            statement.setString(3, item.getPassword());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                item.setId(resultSet.getLong(1));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't add user with id = " + item.getId(), e);
```
Comment (KseniiaMakarova) [inline src/main/java/internet/market/dao/jdbc/UserDaoJdbcImpl.java]: if something goes wrong, your user won't get an id

File: src/main/java/internet/market/dao/jdbc/ShoppingCartDaoJdbcImpl.java [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+            PreparedStatement preparedStatement = connection.prepareStatement(query,
+                    Statement.RETURN_GENERATED_KEYS);
+            preparedStatement.setLong(1, item.getUserId());
+            preparedStatement.executeUpdate();
+            ResultSet resultSet = preparedStatement.getGeneratedKeys();
+            if (resultSet.next()) {
+                item.setId(resultSet.getLong(1));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't add shopping cart with id = "
```
Comment (KseniiaMakarova) [inline src/main/java/internet/market/dao/jdbc/ShoppingCartDaoJdbcImpl.java]: same here

File: src/main/java/internet/market/dao/jdbc/ShoppingCartDaoJdbcImpl.java [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+    @Override
+    public ShoppingCart update(ShoppingCart item) {
+        String query = "UPDATE shopping_carts SET user_id = ? "
+                + "WHERE shopping_cart_id = ? AND deleted = false";
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, item.getUserId());
+            statement.setLong(2, item.getId());
+            statement.executeUpdate();
+            List<Product> allProducts = getProductsFromCart(item.getId());
```
Comment (KseniiaMakarova) [inline src/main/java/internet/market/dao/jdbc/ShoppingCartDaoJdbcImpl.java]: in this method you receive a shopping cart with already updated product list

there is no need to get the products from DB

File: src/main/java/internet/market/dao/jdbc/ShoppingCartDaoJdbcImpl.java [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+            deleteProductsFromCart(item.getId());
+            addProductsToShoppingCart(allProducts, item.getId());
```
Comment (KseniiaMakarova) [inline src/main/java/internet/market/dao/jdbc/ShoppingCartDaoJdbcImpl.java]: both these two methods open connections inside them
let's avoid the situation where we have a connection inside another connection

File: src/main/java/internet/market/dao/jdbc/ShoppingCartDaoJdbcImpl.java [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get products from "
+                    + "shoppingCart with id = " + id, e);
+        }
+        return productList;
+    }
+
+    private void addProductsToShoppingCart(List<Product> products,
+                                           Long cartId) {
```
Comment (KseniiaMakarova) [inline src/main/java/internet/market/dao/jdbc/ShoppingCartDaoJdbcImpl.java:170]:
```suggestion
    private void addProductsToShoppingCart(ShoppingCart cart) {
```

File: src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java (hunk 2 above)
Comment (KseniiaMakarova) [reply src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java:156]: @KostiantynVorobiov пересмотри внимательнее, тут нет такой ситуации

File: src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+            PreparedStatement statement = connection.prepareStatement(query,
+                    Statement.RETURN_GENERATED_KEYS);
+            statement.setLong(1, item.getUserId());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                item.setId(resultSet.getLong(1));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can`t add order with id =  " + item.getId(), e);
```
Comment (KseniiaMakarova) [inline src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java]: same comment about the id as in previous two cases

File: src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+    public Optional<Order> get(Long id) {
+        String query = "SELECT * FROM orders WHERE order_id = ? AND deleted = false";
+        Order order = null;
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                order = getOrderInformationFromDataBase(resultSet);
+                order.setProducts(getProductsOfOrder(id));
```
Comment (KseniiaMakarova) [inline src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java]: you are setting the products twice in this method

File: src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+                return Optional.of(order);
```
Comment (KseniiaMakarova) [inline src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java]: do we need to return here if we decide to remove the previous line?

File: src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+                + "WHERE order_id = ? AND deleted = false";
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement =
+                    connection.prepareStatement(updateOrderQuery);
+            statement.setLong(1, item.getUserId());
+            statement.setLong(2, item.getId());
+            statement.executeUpdate();
+            List<Product> allProducts = getProductsOfOrder(item.getId());
+            deleteProductsFromOrder(item.getId());
+            addProductsToOrder(allProducts, item.getId());
```
Comment (KseniiaMakarova) [inline src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java]: same comment as for shoppingCart `update()` method

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/internet/market/dao/jdbc/ShoppingCartDaoJdbcImpl.java (hunk 10 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/ShoppingCartDaoJdbcImpl.java]: done

File: src/main/java/internet/market/dao/jdbc/UserDaoJdbcImpl.java (hunk 9 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/UserDaoJdbcImpl.java]: done

File: src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java (hunk 14 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java]: done

File: src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java (hunk 15 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java]: done

File: src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java (hunk 16 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java]: done

File: src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java (hunk 17 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java]: done

File: src/main/java/internet/market/dao/jdbc/ShoppingCartDaoJdbcImpl.java (hunk 11 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/ShoppingCartDaoJdbcImpl.java]: done

File: src/main/java/internet/market/dao/jdbc/ShoppingCartDaoJdbcImpl.java (hunk 12 above)
Comment (voloshynamaryna11) [reply src/main/java/internet/market/dao/jdbc/ShoppingCartDaoJdbcImpl.java]: done

File: src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java (hunk 2 above)
Comment (KostiantynVorobiov) [reply src/main/java/internet/market/dao/jdbc/OrderDaoJdbcImpl.java:156]: Я глубоко извеняюсь, недосмотрел

**Pull request merged**